File: website/db/mysql2schema.php
<?php
ini_set('include_path',ini_get('include_path').":./db");
require_once('abstract-mysql.php');
require_once('abstract-common.php');

$m2sdb = new Entity(array(
	'host' => $dbhost,
	'login' => $myuser,
	'pw' => $mypw,
	'db' => $db,
));

$m2sdb->resultformat = 'array';
$m2sdb->run("show tables");
$tables = $m2sdb->resultarray();

foreach ($tables as $row) {
	
	$table = $row[0];

	$m2sdb->run("desc $table");
	$fields = $m2sdb->resultarray();

	$keys = array();
	$tdata = array();
	foreach ($fields as $fdata) {
		$s = array();
		if (preg_match('#(\w+)\((\d+)\)#i', $fdata['Type'],$m)) {
			$s['type']  = strtolower($m[1]);
			$s['size'] = $m[2];
		} else {
			$s['type'] = strtolower($fdata['Type']);
		}
		if ($fdata['Extra'] == 'auto_increment') {
			$s['auto'] = true;
		}
		if ($fdata['Key'] == 'PRI') {
			$s['key'] = true;
			$keys[$fdata['Field']] = '';
		}
		$s['null'] = ($fdata['Null'] == 'YES' ? true: false);
		$s['default'] = $fdata['Default'];
		$s['name'] = $fdata['Field'];
		$s['extra'] = $fdata['Extra'];
		$tdata[$s['name']] = $s;
	}
	if (count(array_keys($keys)) != 1) {
		$tdata['PRIMARY KEY'] = $keys;
	}
	$schema[$table] = $tdata;
}

$schemavar = var_export($schema,true);
$schemaphp = <<<PHP
<?php
\$dbhost = '$dbhost';
\$dbname = '$db';
\$schema = $schemavar;
PHP;

if ($schemafile) {
	$fh = fopen($schemafile,'wb');
	if (!$fh) die("can't open $schemafile");
	$ret = fwrite($fh,$schemaphp);
	fclose($fh);
	if (!$ret) die("error writing to $schemafile");
} else {
	print $schemaphp;
}

File: website/db/mysql2schema.pl
#!/usr/bin/perl
# translate a mysql schema into something that can be made into a set of class defs
use strict;
my (@fields, %schema, @keys);

print "<?php\n\$schema = array();\n\n";

while (<>) {
	# this will be hidden in a comment at the top of the dump
	if (/Host:\s*(\S*)\s*Database:\s*(\S*)/) {
		print "\$dbhost = '$1';\n\$dbname = '$2';\n\n";
		next;
	}
	if (/^CREATE TABLE/.../^\).*;$/) {
		if (/^CREATE TABLE `([^`]+)`/) {
			my $table = $1;
			if (scalar(@fields)) { &printschema(); }
			print "\$schema['$table'] = array(\n";

		} elsif (/^\s*`([^`]*)` (\w+)(?:\((\d*)\)|(\(.*\))|)/) {
			my ($field,$type,$size,$opts) = ($1,$2,$3,$4);
			if ($field =~ /date|time/) {
				$size = 20;
			}
			$schema{$field} =  { type => "'$type'", size => $size };
			$schema{$field}{auto} = 1 if /AUTO_INCREMENT/i;
			if ($type =~ /enum/i) {
				$schema{$field}{opts} = "array$opts";
			}
			push @fields, $field;

		} elsif (/^\s*PRIMARY KEY\s+\((.*)\)/) {
			my $keys = $1;
			$keys =~ s/`//g;
			@keys = split ',', $keys;
			if (scalar(@keys) == 1) {
				foreach my $field (keys %schema) {
					$keys =~ s/\(\d+\)$//;
					if ($field =~ m/^($keys)$/) {
						$schema{$field}{key} = 'true';
					}
				}
			} else {
				print "\t'PRIMARY KEY' => array(";
				foreach my $key (@keys) {
					my $tb;
					$key =~ s/\(.*//;
					if ($key =~ /(\w+)_id/) {
						$tb = $1;
					}
					print "'$key' => '$tb', ";
				}
				print "),\n";
			}
		}	
	}
}

&printschema if scalar @fields;

sub printschema {
	foreach my $field (@fields) {
		print "\t'$field' => array( ";
		foreach my $prop (qw/type size rows cols key auto opts/) {
			next unless defined $schema{$field}{$prop};
			print "'$prop' => $schema{$field}{$prop}, ";
		}
		print "),\n";
	}
	print ");\n\n";
	%schema = ();
	@fields = ();
}
__END__
=head1 Name

mysql2schema.pl

=head1 Description 

purpose of this script is to read mysql CREATE TABLE statements and 
make a schema file from them 

=head1 Synopsis

 mysql2schema.pl mysql-dump-file > my-schema-file.php
 
or 

 mysqldump ... | mysql2schema.pl > my-schema-file.php

you should then edit the my-schema-file.php file as needed before making 
Entity and Relation child classes based on each table

=head1 TODO

make a tool to do this automatically

=head1 Notes

if you use the form {tablename}_id for the primary key for Entity tables
(ie where there is only one key) this will correctly add the table
but this can be added by hand later

=head1 Copyright

 Author Cal Woodruff cwoodruf@gmail.com
 Licensed under the Perl Artistic License version 2.0
 http://www.perlfoundation.org/attachment/legal/artistic-2_0.txt

=cut
